dataflow/snippets/src/main/java/com/example/dataflow/KafkaRead.java

```diff
@@ -0,0 +1,91 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.dataflow;
+
+// [START dataflow_kafka_read]
+import com.google.common.collect.ImmutableMap;
+import java.io.UnsupportedEncodingException;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.TextIO;
+import org.apache.beam.sdk.managed.Managed;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.options.StreamingOptions;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+public class KafkaRead {
+
+  public interface Options extends StreamingOptions {
+    @Description("The Kafka bootstrap server. Example: localhost:9092")
+    String getBootstrapServer();
+
+    void setBootstrapServer(String value);
+
+    @Description("The Kafka topic to read from.")
+    String getTopic();
+
+    void setTopic(String value);
+
+    @Description("Path to write the output file")
+    String getOutputPath();
+
+    void setOutputPath(String value);
+  }
+
+  public static void main(String[] args) {
+    // Parse the pipeline options passed into the application. Example:
+    //   --bootstrap_servers=$BOOTSTRAP_SERVERS --topic=$KAFKA_TOPIC --outputPath=$OUTPUT_FILE
+    // For more information, see https://beam.apache.org/documentation/programming-guide/#configuring-pipeline-options
+    var options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
+    options.setStreaming(true);
+
+    var pipeline = Pipeline.create(options);
+    ImmutableMap<String, Object> config = ImmutableMap.<String, Object>builder()
+        .put("bootstrap_servers", options.getBootstrapServer())
+        .put("topic", options.getTopic())
+        .put("data_format", "RAW")
+        .put("max_read_time_seconds", 15)
+        .put("auto_offset_reset_config", "earliest")
+        .build();
+
+    // Build the pipeline.
+    pipeline
+        // Read messages from Kafka.
+        .apply(Managed.read(Managed.KAFKA).withConfig(config)).getSinglePCollection()
+        // Get the payload of each message and convert to a string.
+        .apply(MapElements
+            .into(TypeDescriptors.strings())
+            .via((row -> {
+              var bytes = row.getBytes("payload");
+              try {
+                return new String(bytes, "UTF-8");
+              } catch (UnsupportedEncodingException e) {
+                throw new RuntimeException(e);
+              }
+            })))
+        // Write the payload to a text file.
+        .apply(TextIO
+            .write()
+            .to(options.getOutputPath())
+            .withSuffix(".txt")
+            .withNumShards(1)
+    );
+    pipeline.run().waitUntilFinish();
+  }
+}
```

Comment:
please move the code sample code to a stand-alone method.
call this method from the main pipeline.
implement a simple response in the method.
consider returning a response that can be easily validated in testing.

Reply:
I could move the pipeline code into a separate method, but I need to include the Main method in the snippet. The Beam runner (which includes the Dataflow runner) invokes the Main method, so that's part of the sample.

Building the pipeline in Main is idiomatic for Beam: https://github.com/apache/beam/blob/master/examples/java/src/main/java/org/apache/beam/examples/MinimalWordCount.java

I'll update to return the pipeline result.

Reply:
I understand. We distinguish between the code sample and code snippet. The latter is displayed in the documentation while the former is a fully compilable and executable code.
Unless there is something specific about the code in the `main()` method of _any_ Java application that you want to demonstrate, the startup code (i.e. `main` method and other auxiliary lines) should be excluded from the snippet. This way the documentation shows only the lines that are relevant in the context and a reader get access the full code sample via a link to Github provided in the snippet widget.

Reply:
I trimmed out main() by moving the logic for running the pipeline into the test, also the logic for instantiating the options class. That leaves the snippet focused on building the pipeline.

re: the options class, See my other comment. I do need to keep this class for the snippet.